poincare/src/evaluation.cpp

```diff
@@ -0,0 +1,261 @@
+extern "C" {
+#include <assert.h>
+#include <float.h>
+#include <stdlib.h>
+}
+#include <poincare/evaluation.h>
+#include <poincare/expression.h>
+#include <cmath>
+
+namespace Poincare {
+
+template<typename T>
+T Complex<T>::toScalar() const {
+  if (this->imag() == 0.0) {
+    return this->real();
+  }
+  return NAN;
+}
+
+template<typename T>
+std::complex<T> Complex<T>::pow(const std::complex<T> &c, const std::complex<T> &d) {
+  std::complex<T> result = std::pow(c, d);
+  /* Cheat: pow openbsd immplementationd is a numerical approximation.
+   * We though want to avoid evaluating e^(i*pi) to -1+1E-17i. We thus round
+   * real and imaginary parts to 0 if they are negligible compared to their
+   * norm. */
+  T norm = std::norm(result);
+  if (norm !=  0 && std::fabs(result.real()/norm) <= Expression::epsilon<T>()) {
+    result.real(0);
+  }
+  if (norm !=  0 && std::fabs(result.imag()/norm) <= Expression::epsilon<T>()) {
+    result.imag(0);
+  }
+  if (c.imag() == 0 && d.imag() == 0 && (c.real() >= 0 || d.real() == std::round(d.real()))) {
+    result.imag(0);
+  }
+  return result;
+}
+
+template<typename T>
+std::complex<T> Complex<T>::approximate(const std::complex<T>& c, ComplexFunction<T> function) {
```

Comment:
* Remove `approximate` from `Evaluation`
* Remove the `sin`, `cos` et al from `Evaluation` as well
* In Sine:
```
static std::complex<T> Sine::computeOnComplex(std::complex<T> c) {
  std::complex res = std::sin(c);
  return Trigonometry::roundToMeaningfulDigits(res);
}
```
* In Power:
```
static std::complex<T> Sine::computeOnComplex(std::complex<T> c) {
 // Here, the code that was in `Evaluation`
}
```